### PR TITLE
Feature/MUNI Proxy

### PR DESCRIPTION
**Pull request by chanharryhk**
Decoupled MUNI into MUNI Logic and MUNI State for upgradability

### CODE REVIEW

**Comment by chanharryhk** [comment]
One of the concerns that Kyle and I had was removing the `immutable` keyword from the pool variable in MUNI.sol. Since variables are no longer initialized in the constructor we shouldn't use `immutable` anymore. There is currently no logic implemented to change the pool variable so thats good but just wanted your thoughts.

**Comment by kendricktan** [comment]
Yes you *should* remove the immutable keyword when using proxies as immutable variables will write it to the (proxy's) bytecode making it non-changable

**Review by kendricktan — CHANGES_REQUESTED** [review]
Looks good overall!

To ensure it really is upgradable I would try make a new logic contract in the test with a different func/name and upgrade the logic to it, call it and see if the logic did changed

File: src/test/liquidity-mining/UniswapV3/MUNINewLogic.sol [hunk 1]
```diff
@@ -0,0 +1,885 @@
+// SPDX-License-Identifier: UNLICENSED
```
Comment (kendricktan) [inline src/test/liquidity-mining/UniswapV3/MUNINewLogic.sol:1]: Do we really need a new file here? Would probably be easier to inherit from the original logic contract and add a new function (e.g. `function version() public returns (uint256) { returns 1; }`)), try and call that function and see if its successful?

**Review by kendricktan — CHANGES_REQUESTED** [review]
I would recommend making MUNINewLogic.sol inherit from the existing logic and add a new / override an existing function

**Comment by kafitz** [comment]
One other looming question I had is about a re-initialization check here: https://github.com/dfx-finance/asc/blob/a97d44c7b87525766cde5a720d9007e0e71099de/src/liquidity-mining/UniswapV3/MUNILogicV1.sol#L58-L97

There original MUNI.sol had none with the note "We don't need to check twice if the contract's initialized as __ERC20_init does that check". However, MUNI.sol's constructor looks like ` constructor() ERC20(_name, _symbol) {}` whereas our upgradeable version was replaced with `function initialize( ) public initializer {}`. Either my `isInitialized` guard can go or the note about __ERC20_init handling it should be deleted.

**Comment by kafitz** [comment]
@chanharryhk I created a refactored version here. I didn't strictly follow Kendrick's advice since now I have a new file in both directories, but I thought it would be clean to do the upgrade in its own test. Feel free to reject this branch.

- https://github.com/dfx-finance/asc/tree/feature/muni-proxy-upgrade-test-refactor/src/test/liquidity-mining/UniswapV3
- https://github.com/dfx-finance/asc/blob/feature/muni-proxy-upgrade-test-refactor/src/liquidity-mining/UniswapV3/MUNINewLogic.sol

Also I wasn't sure, is it required to copy/paste all the previous methods each time or is inheritance enough? Harry's tests seem to pass fine here reading the old state and a new method from an upgraded contract. A lot of the docs out there talk about redeclaring everything, but this blog seemed to show you could just inherit and add a method ¯\_(ツ)_/¯ https://dev.to/yakult/tutorial-write-upgradeable-smart-contract-proxy-contract-with-openzeppelin-1916

**Comment by kendricktan** [comment]
Lets remove `__ERC20_init`, seems like its not doing an initializer check.

> is it required to copy/paste all the previous methods each time or is inheritance enough

Its not, you should be able to just inherit from the old logic and add a new (pure/view-only function) that *only* works for that version. If the call to the function that only exists in the later version fails when you've upgraded the contract then the upgrade failed.

IMO we shouldn't be copy+pasting so much code around unless its absolutely necessary as it complicates the whole structure.

@kafitz Thats perfect! Could you move https://github.com/dfx-finance/asc/blob/feature/muni-proxy-upgrade-test-refactor/src/liquidity-mining/UniswapV3/MUNINewLogic.sol into the current branch? And also change the file location to tests since that file isn't really used anywhere apart from tests

**Review by kendricktan — APPROVED** [review]
LGTM

**Pull request merged**